File: backtracking_graph_coloring.cpp
#include <iostream>
#include <vector>

using namespace std;

#define V 4 // Total Vertices

bool isSafe(int v, vector<vector<int>>& graph, vector<int>& color, int c) {
    for (int i = 0; i < V; i++) {
        if (graph[v][i] && c == color[i]) return false;
    }
    return true;
}

bool graphColoringUtil(vector<vector<int>>& graph, int m, vector<int>& color, int v) {
    if (v == V) return true;

    for (int c = 1; c <= m; c++) {
        if (isSafe(v, graph, color, c)) {
            color[v] = c;
            if (graphColoringUtil(graph, m, color, v + 1)) return true;
            color[v] = 0; // Backtrack
        }
    }
    return false;
}

int main() {
    cout << "--- DSA Masterclass: Graph Coloring (C++) ---" << endl << endl;

    vector<vector<int>> graph = {
        {0, 1, 1, 1},
        {1, 0, 1, 0},
        {1, 1, 0, 1},
        {1, 0, 1, 0}
    };
    int m = 3; // Colors available
    vector<int> color(V, 0);

    if (!graphColoringUtil(graph, m, color, 0)) {
        cout << "Solution does not exist" << endl;
    } else {
        cout << "Graph Vertex Coloring Scheme (C++):" << endl;
        for (int i = 0; i < V; i++) {
            cout << " -> Vertex " << i << " -> Color ID: " << color[i] << endl;
        }
    }
    return 0;
}

File: backtracking_graph_coloring.py
def is_safe(v, graph, color, c, num_vertices):
    # Check if any adjacent vertex has the same color
    for i in range(num_vertices):
        if graph[v][i] == 1 and color[i] == c:
            return False
    return True

def graph_coloring_util(graph, m, color, v, num_vertices):
    # Base Case: All vertices are colored
    if v == num_vertices:
        return True

    # Try different colors for vertex v
    for c in range(1, m + 1):
        if is_safe(v, graph, color, c, num_vertices):
            color[v] = c
            
            # Recur to color rest of the vertices
            if graph_coloring_util(graph, m, color, v + 1, num_vertices):
                return True
            
            # BACKTRACK
            color[v] = 0
            
    return False

def main():
    print("--- DSA Masterclass: Graph Coloring (Python) ---\n")
    
    # Adjacency Matrix representation
    graph = [
        [0, 1, 1, 1],
        [1, 0, 1, 0],
        [1, 1, 0, 1],
        [1, 0, 1, 0]
    ]
    
    num_vertices = len(graph)
    m = 3  # Number of available colors
    color = [0] * num_vertices  # Initialize colors as 0

    if not graph_coloring_util(graph, m, color, 0, num_vertices):
        print("Solution does not exist")
    else:
        print("Graph Vertex Coloring Scheme (Python):")
        for i in range(num_vertices):
            print(f" -> Vertex {i} -> Color ID: {color[i]}")

if __name__ == "__main__":
    main()